core/app/models/spree/payment/processing.rb

```diff
@@ -139,77 +164,56 @@ def handle_void_response(response)
 
       private
 
-      def process_authorization
-        started_processing!
-        gateway_action(source, :authorize, :pend)
-      end
+      # @raises Spree::Core::GatewayError
+      def check_payment_preconditions!
+        return if processing?
+        return unless payment_method
+        return unless payment_method.source_required?
 
-      def process_purchase
-        started_processing!
-        gateway_action(source, :purchase, :complete)
-        # This won't be called if gateway_action raises a GatewayError
-        capture_events.create!(amount: amount)
-      end
-
-      def handle_payment_preconditions(&_block)
-        unless block_given?
-          raise ArgumentError.new("handle_payment_preconditions must be called with a block")
+        unless source
+          gateway_error(I18n.t('spree.payment_processing_failed'))
         end
 
-        return if payment_method.nil?
-        return if !payment_method.source_required?
-
-        if source
-          if !processing?
-            if payment_method.supports?(source)
-              yield
-            else
-              invalidate!
-              raise Core::GatewayError.new(I18n.t('spree.payment_method_not_supported'))
-            end
-          end
-        else
-          raise Core::GatewayError.new(I18n.t('spree.payment_processing_failed'))
+        unless payment_method.supports?(source)
+          invalidate!
+          gateway_error(I18n.t('spree.payment_method_not_supported'))
         end
-      end
 
-      def gateway_action(source, action, success_state)
```

Comment:
Found an occurrence of this here: https://github.com/solidusio/solidus_paypal_braintree/blob/806d20b50fbb2af6762b3dabc6a210306fa7d89f/spec/models/solidus_paypal_braintree/gateway_spec.rb#L103. Maybe we can change the spec first to stop relying on a private method if possible?

Reply:
Good finding!

Luckily this looks like a false positive, `gateway_action` in that contexts refers to an RSpec `let` that points to a method of `Braintree::TransactionGateway`.